src/cargo/sources/config.rs

```diff
@@ -81,6 +84,22 @@ impl<'gctx> SourceConfigMap<'gctx> {
                 base.add_config(key, value)?;
             }
         }
+
+        for (key, path) in gctx.local_overlays().iter() {
+            let overlay_id = SourceId::for_local_registry(path)?;
+            base.overlays.insert(*key, overlay_id);
+        }
+
+        if let Ok(overlay) =
+            gctx.get_env("__CARGO_TEST_DEPENDENCY_CONFUSION_VULNERABILITY_DO_NOT_USE_THIS")
+        {
+            let (url, path) = overlay.split_once('=').ok_or(anyhow::anyhow!(
+                "invalid overlay format. I won't tell you why; you shouldn't be using it anyway"
+            ))?;
+            let overlay_id = SourceId::for_local_registry(path.as_ref())?;
+            base.overlays.insert(SourceId::from_url(url)?, overlay_id);
+        }
```

Comment:
Should we move this hack up into the caller?  `PackageRegistry` is the immediate caller but maybe a layer above that?  That would help with the `GlobalContext`.  The main downside is it would make it more annoying to support `__CARGO_TEST_DEPENDENCY_CONFUSION_VULNERABILITY_DO_NOT_USE_THIS`

Thoughts?

Reply:
I think if we moved the overlay configuration into `Workspace` as discussed below then this could be pushed up into `resolve_ws` and `resolve_ws_with_opts`

Reply:
Just wanted to check in if there is anything you are waiting on me for this change as you've done other work on this PR but not this part yet.

Reply:
Ah sorry, I thought this would go in a follow-up (along with moving the overlay configuration to `Workspace`). On Monday I'll clean up the commit history and look into doing this too.